### PR TITLE
[JENKINS-45047] Support for plugin-to-plugin integration tests

### PR DESCRIPTION
**Pull request by basil**
Finishes the great work started by @jglick in #66. I have left extensive comments in the implementation, so I will not write a long description here. This changes goes along with https://github.com/jenkinsci/plugin-compat-tester/pull/360 and https://github.com/jenkinsci/bom/pull/1059, which are getting close to ready but not quite. This change has been tested enough by now that I think it is ready for review. I have tested this PR in conjunction with the other two and tests have passed successfully on 100 out of 101 plugins in the managed set in `bom-weekly` (there is still a problem with `checks-api`, but I think it is a technical problem with the plugin itself and not with this code). In the coming days I will try this on `bom-2.289.x` but I fear some of the plugins in that line may be too old to support a newer toolchain.

Closes #66.

### CODE REVIEW

**Review by jglick — APPROVED** [review]
Very nice! Thank you for finally moving this forward. Seems safe to merge at any time and iterate, since it should have no effect when the new properties are not being used.

> I have tested this PR in conjunction with the other two and tests have passed successfully on…

To what extent can we verify that in the `bom` run the test classpaths are actually what we expect without the POM rewrites? I suppose it would be compelling to find a recent example of a plugin update which is _known_ to introduce a downstream regression?

File: src/main/java/org/jenkinsci/maven/plugins/hpi/TestDependencyMojo.java
Comment (basil) [reply src/main/java/org/jenkinsci/maven/plugins/hpi/TestDependencyMojo.java:673]: Exactly. Stop telling me I can't code, SpotBugs!

File: src/main/java/org/jenkinsci/maven/plugins/hpi/TestDependencyMojo.java
Comment (basil) [reply src/main/java/org/jenkinsci/maven/plugins/hpi/TestDependencyMojo.java:206]: Yes, exactly. You can see this in action by finding a plugin with an older core baseline (e.g., `text-finder`) and testing with this PR and a megawar of a very new core. The deletions list is a nice summary of all the old JARs I have been removing from core over the past year.

File: src/main/java/org/jenkinsci/maven/plugins/hpi/TestDependencyMojo.java
Comment (basil) [reply src/main/java/org/jenkinsci/maven/plugins/hpi/TestDependencyMojo.java:421]: That last sentence was specifically referring to updating the POM with the results of upper bounds analysis: something which could possibly be automated but would require a lot of fragile parsing of Maven output, munging of the POM with the results, etc. Difficult to maintain, I think.

File: src/main/java/org/jenkinsci/maven/plugins/hpi/TestDependencyMojo.java
Comment (basil) [reply src/main/java/org/jenkinsci/maven/plugins/hpi/TestDependencyMojo.java:458]: Yeah, that and the application of upper bounds analysis. While it might be technically feasible to munge the XML file from the POM-rewriting part of PCT with the right algorithm, it isn't possible to determine upper bounds without actual dependency resolution.

File: src/main/java/org/jenkinsci/maven/plugins/hpi/TestDependencyMojo.java
Comment (basil) [reply src/main/java/org/jenkinsci/maven/plugins/hpi/TestDependencyMojo.java]: Yes, thank you. I improved the comment in commit 6233896.

File: src/main/java/org/jenkinsci/maven/plugins/hpi/TestDependencyMojo.java
Comment (basil) [reply src/main/java/org/jenkinsci/maven/plugins/hpi/TestDependencyMojo.java]: Copypasta, my bad. Fixed in commit cb9b5bb.

File: src/main/java/org/jenkinsci/maven/plugins/hpi/TestDependencyMojo.java
Comment (basil) [reply src/main/java/org/jenkinsci/maven/plugins/hpi/TestDependencyMojo.java]: Done in commit 353da66.

File: src/main/java/org/jenkinsci/maven/plugins/hpi/TestDependencyMojo.java
Comment (basil) [reply src/main/java/org/jenkinsci/maven/plugins/hpi/TestDependencyMojo.java:384]: > I think you are talking about the core BOM here? Took me a while to read this but I think it makes sense.

Yes, exactly. Took me a while to figure out too. Unfortunate but I think there is no way around this given the existing design of the core BOM and plugin POM.

> (Should be noted in `overrideWar` docs.)

Done in commit 7842644.

**Comment by basil** [comment]
Thanks for the review Jesse.

> To what extent can we verify that in the bom run the test classpaths are actually what we expect without the POM rewrites?

Not easily. But essentially the algorithm in this PR grew out of what I did manually dozens of times during the Guava upgrade project, and I spent while ~developing~ debugging this code I spent quite a bit of time comparing, side-by-side, the results of automation with the results of manual POM editing for a number of plugins, so I am feeling confident that the manipulation is working as expected overall. The only thing I have not attempted to deal with at all is split plugins, but neither does `jenkinsci/bom`, so that can be done in some future enhancement if needed.

**Comment by basil** [comment]
I should also add that in this PR I'm printing the final dependency tree by invoking the `tree` goal from the dependency plugin against the synthetic dependency tree, which I found essential while debugging. So the process for debugging is pretty straightforward: apply the expected changes to a POM manually, run `mvn dependency:tree`, and compare the results to the automation. They should be identical, and if they are not the difference should point to where the bug is.

**Comment by basil** [comment]
(This `tree` debugging logic was written in order for me to be able to debug the managed/provided core BOM issue I had, which is what provoked the `jenkins.version` logic and related comment.)

File: src/main/java/org/jenkinsci/maven/plugins/hpi/TestDependencyMojo.java
Comment (jglick) [reply src/main/java/org/jenkinsci/maven/plugins/hpi/TestDependencyMojo.java:206]: Oh this applies to the `jenkins-core` dep too. Interesting, though if you are required to pass `-Djenkins.version=…` anyway, would that not already be making that change to the original project model?

Comment (basil) [reply src/main/java/org/jenkinsci/maven/plugins/hpi/TestDependencyMojo.java:206]: > Oh this applies to the `jenkins-core` dep too. Interesting, though if you are required to pass `-Djenkins.version=…` anyway, would that not already be making that change to the original project model?

It would... now that I think about it, I went through that particular debugging exercise before I had added the requirement for `-Djenkins.version`, so I guess you wouldn't be able to replicate it now. :smile:

**Comment by jglick** [comment]
> verify that in the bom run the test classpaths are actually what we expect

I realize after I wrote this that I was actually asking something much more modest than the phrasing implied: a sanity check that a sampling of a few representative cases does show tests being run against _some_ plugin versions taken from the BOM under test rather than the checked-in POM. For example, if there is a confirmed reproduction case for https://github.com/jenkinsci/bom/issues/821, that using this new mode indeed appears to fix it; or test output somewhere in https://github.com/jenkinsci/bom/pull/1059 that offers a clear example of a test running using a newer plugin dep version than what `pct-work/*/pom.xml` requests.

**Review by jglick — APPROVED** [review]

**Comment by basil** [comment]
> or test output somewhere in [jenkinsci/bom#1059](https://github.com/jenkinsci/bom/pull/1059) that offers a clear example of a test running using a newer plugin dep version than what `pct-work/*/pom.xml` requests.

I checked the `-cp` argument to the Surefire JVM numerous times while developing this change and I am sure that it is working Jesse.

**Comment by basil** [comment]
I started testing `bom-weekly` with upper bounds analysis and found some interesting issues. Here's an interesting upper bounds failure. From e.g. `caffeine-api` try

```
$ diff --git a/pom.xml b/pom.xml
diff --git a/pom.xml b/pom.xml
index 3f2a8d4..d3b0b17 100644
--- a/pom.xml
+++ b/pom.xml
@@ -56,6 +56,18 @@
         </exclusion>
       </exclusions>
     </dependency>
+    <dependency>
+        <groupId>io.jenkins.plugins</groupId>
+        <artifactId>bootstrap4-api</artifactId>
+        <version>4.6.0-4</version>
+        <scope>test</scope>
+    </dependency>
+    <dependency>
+        <groupId>io.jenkins.plugins</groupId>
+        <artifactId>popper-api</artifactId>
+        <version>1.16.1-3</version>
+        <scope>test</scope>
+    </dependency>
   </dependencies>
```

That is designed to simulate a scenario where you have both plugins installed on a real Jenkins controller. This results in:

```
Require upper bound dependencies error for org.webjars:popper.js:1.16.1-lts [test] paths to dependency are:
+-io.jenkins.plugins:caffeine-api:2.9.3-0-SNAPSHOT
  +-io.jenkins.plugins:popper-api:1.16.1-3 [test]
    +-org.webjars:popper.js:1.16.1-lts [test]
and
+-io.jenkins.plugins:caffeine-api:2.9.3-0-SNAPSHOT
  +-io.jenkins.plugins:bootstrap4-api:4.6.0-4 [test]
    +-org.webjars:bootstrap:4.6.0 [test]
      +-org.webjars:popper.js:2.5.4 [test]
```

From `bootstrap4-api`'s tree:

```
[INFO] io.jenkins.plugins:bootstrap4-api:hpi:4.6.0-5-SNAPSHOT
[INFO] +- org.webjars:bootstrap:jar:4.6.0:provided
[INFO] |  \- org.webjars:popper.js:jar:2.5.4:provided
[INFO] +- io.jenkins.plugins:popper-api:jar:1.16.1-2:compile
```

This looks odd to me. Does Bootstrap 4 depend on Popper v1 or Popper v2? If v1, then what is `org.webjars:popper.js:jar:2.5.4` doing as a dependency of `org.webjars:bootstrap:jar:4.6.0`? If v2, then why declare a plugin-to-plugin dependency on `io.jenkins.plugins:popper-api:jar:1.16.1-2`? Either way something is not adding up to me. Perhaps @uhafner can shed some light on this.

File: src/main/java/org/jenkinsci/maven/plugins/hpi/TestDependencyMojo.java
Comment (jtnord) [inline src/main/java/org/jenkinsci/maven/plugins/hpi/TestDependencyMojo.java:237]: should this not be only doing this if the version is newer (using a maven version comparison) and presumably aborting if the war contains an older version of the dependency?
(in other words the prepared war may have all the latest releases of a plugin, but if you are running this on a plugin that has a dependency on an incremenatal the results will be incorrect and this user error should be reported rather than ignored?)

File: src/main/java/org/jenkinsci/maven/plugins/hpi/TestDependencyMojo.java
Comment (jtnord) [inline src/main/java/org/jenkinsci/maven/plugins/hpi/TestDependencyMojo.java:263]: NIT: I would expect this to be at info level and the dependency:tree to be debug.

File: src/main/java/org/jenkinsci/maven/plugins/hpi/TestDependencyMojo.java
Comment (jtnord) [inline src/main/java/org/jenkinsci/maven/plugins/hpi/TestDependencyMojo.java:361]: Patterns are thread safe and compilation is not free so should not be recompiled in loops.

```suggestion
            final Pattern corePattern = Pattern.compile(CORE_REGEX);
            final Pattern pluginPattern = Pattern.compile(PLUGIN_REGEX);
            while (entries.hasMoreElements()) {
```

File: src/main/java/org/jenkinsci/maven/plugins/hpi/TestDependencyMojo.java
Comment (jtnord) [inline src/main/java/org/jenkinsci/maven/plugins/hpi/TestDependencyMojo.java:371]:
```suggestion
                m = pluginPattern.matcher(name);
```

**Review by jtnord — COMMENTED** [review]
looks good.

I do not follow the issue with the BOM though - it seems like we can add a `dependencyManagement` for the jenkins bom to the project in order to resolve the correct core libraries without having to pass `jenkins.version` and then we can have a single step "PCT" IIUC.  e

File: src/main/java/org/jenkinsci/maven/plugins/hpi/TestDependencyMojo.java
Comment (basil) [reply src/main/java/org/jenkinsci/maven/plugins/hpi/TestDependencyMojo.java:361]: This code was copied from `plugin-compat-tester`. I note that you are comfortable pointing out the flaw in the context of this PR but not comfortable opening a pull request to fix the issue in `plugin-compat-tester`.

File: src/main/java/org/jenkinsci/maven/plugins/hpi/TestDependencyMojo.java
Comment (basil) [reply src/main/java/org/jenkinsci/maven/plugins/hpi/TestDependencyMojo.java:264]: For what reason do you have this expectation?

File: src/main/java/org/jenkinsci/maven/plugins/hpi/TestDependencyMojo.java
Comment (basil) [reply src/main/java/org/jenkinsci/maven/plugins/hpi/TestDependencyMojo.java:162]: I am sure we could, James. I did not implement this because I could not think of an easy way to test it.

File: src/main/java/org/jenkinsci/maven/plugins/hpi/TestDependencyMojo.java
Comment (basil) [reply src/main/java/org/jenkinsci/maven/plugins/hpi/TestDependencyMojo.java:255]: I think the argument about it being too verbose makes sense. I swapped them in commit 2ca309b.

File: src/main/java/org/jenkinsci/maven/plugins/hpi/TestDependencyMojo.java
Comment (basil) [reply src/main/java/org/jenkinsci/maven/plugins/hpi/TestDependencyMojo.java:238]: I thought so too originally, but that line of reasoning turned out to be flawed. The situation is as follows: suppose we depend on A:1.0 which depends on B:1.2. Now suppose a problem is discovered in B:1.2 that results in A:1.1 rolling back to B:1.1. We only ever directly depended on A:1.0, but now we override A:1.0 to A:1.1. B:1.2 was in our transitive tree before, but now for correctness we must change B from 1.2 to 1.1.

Comment (jtnord) [reply src/main/java/org/jenkinsci/maven/plugins/hpi/TestDependencyMojo.java:238]: Makes sense thanks for the explanation .  Worth a comment for future contributors?

```suggestion
                    /*
                     * We do an equals check here rather than a new for the following reason: 
                     *  suppose we depend on A:1.0 which depends on B:1.2. Now suppose a problem is discovered in B:1.2
                     * that results in A:1.1 rolling back to B:1.1. We only ever directly depended on A:1.0, but now we
                     *  override A:1.0 to A:1.1. B:1.2 was in our transitive tree before, but now for correctness we must
                     * change B from 1.2 to 1.1.
                     */
                    if (!newVersion.equals(originalVersion)) {
```

File: src/main/java/org/jenkinsci/maven/plugins/hpi/TestDependencyMojo.java
Comment (jtnord) [reply src/main/java/org/jenkinsci/maven/plugins/hpi/TestDependencyMojo.java:362]: I did not know this came from the PCT.

File: src/main/java/org/jenkinsci/maven/plugins/hpi/TestDependencyMojo.java
Comment (basil) [reply src/main/java/org/jenkinsci/maven/plugins/hpi/TestDependencyMojo.java]: Done in commit 937ef21.

**Comment by basil** [comment]
> looks good.

And yet you have not approved the PR.

> I do not follow the issue with the BOM though - it seems like we can add a `dependencyManagement` for the jenkins bom to the project in order to resolve the correct core libraries without having to pass `jenkins.version` and then we can have a single step "PCT" IIUC. e

Thank you for the code review, James. In addition to submitting code reviews, I also encourage you to submit pull requests to `jenkinsci/maven-hpi-plugin`, `jenkinsci/plugin-compat-tester`, and `jenkinsci/bom`.

**Comment by jtnord** [comment]
> I do not follow the issue with the BOM though - it seems like we can add a `dependencyManagement` for the jenkins bom to the project in order to resolve the correct core libraries without having to pass `jenkins.version` and then we can have a single step "PCT" IIUC. e

> Thank you for the code review, James. In addition to submitting code reviews, I also encourage you to submit pull requests to jenkinsci/maven-hpi-plugin, jenkinsci/plugin-compat-tester, and jenkinsci/bom.


my point was I did not follow the conversation / code and due to that it looks wrong to me, and I did not make this clear as it has not been answered. 

So to try again.

It looks like the expectation to have to pass in `jenkins.version` is wrong - we can have the jenkins.verion from the war in the case it is provided, and then can use that to update the jenkins-bom in the depnednecy management.  So there is no need to change the `jenkins.version` and do another phase, as we are manipulating the project model and re-forcing dependency resolution - so why is the bom not like anything else I fundamentally do not understand and neither the review comments nor the code comment help me to understand this (and if I do not understand it in the context of this PR, I am not sure others would in the future either).

And if this expectation is indeed incorrect and we did update the jenkins bom just like everything else then PCT can be nuked from orbit.  We should not even need the PCT for plugins using an older parent as again to my knowledge the hpi-plugin has a stable mojo API.

So to run a "PCT" for a plugin all you need to do now is assuming a new enough parent pom.

```
createUberWarWithAllPLuginsIWantToTestAgainst.sh
git clone the-plugin.git
cd the-plugin
mvn -DsoverrideWar=fat.war test
```

and if the parent pom is not new enough we can do the following actually specify the plugin version on the CLI as it is (in all version 4 of the `plugin-pom` a property)

```
createUberWarWithAllPLuginsIWantToTestAgainst.sh
git clone the-plugin.git
cd the-plugin
mvn -DsoverrideWar=fat.war -Dhpi-plugin.version=3.28 test
```

And then to ignore parsing the POMs to find out what version we have - we can extract the version (or manually update it whenever we release a new version which is not so often) and 

```
createUberWarWithAllPLuginsIWantToTestAgainst.sh
git clone the-plugin.git
cd the-plugin
mvn v
mvn -DsoverrideWar=fat.war -Dhpi-plugin.version=3.28 test
```

But doing the above is dependant on the code in here which I left a comment about which seems to be wrong.

If I am indeed correct - I would be happy to nuke the PCT from orbit and update the BOM build, but perhaps I am missing some tribal knowledge of this area that is not explained in this code or accompanying comments.

> And yet you have not approved the PR.

because I left a review comment where I felt the code was not correct, but as I was unsure I was not going to block the merging.

**Comment by jglick** [comment]
> the expectation to have to pass in `jenkins.version is wrong` - we can [infer that from `overrideWar`], and then can use that to update the `jenkins-bom` in [`dependencyManagement`]

Possibly so, assuming that the somewhat scary tricks here used to recreate a POM model in memory will honor that update. That would save you from the trouble of passing `-Djenkins.version`, and would allow you to test a plugin against a megawar in a single phase rather than the two phases we would otherwise need to prove that we are enforcing binary and not source compatibility. Discussed earlier in https://github.com/jenkinsci/maven-hpi-plugin/pull/336#discussion_r860917906. So

```bash
mvn test-compile
mvn -Djenkins.version=… -DoverrideWar=… hpi:resolve-test-dependencies surefire:test
```

could be simplified to

```bash
mvn -DoverrideWar=… test
```

Seems like it would be a welcome improvement, but I do not think lack of that should block us from moving forward—someone can add it as time permits, unless I am missing something?

> to my knowledge the hpi-plugin has a stable mojo API

For the most part, though we do sometimes introduce breaking changes, like #302. Less of an issue if we are anyway doing a two-phase build, since then the plugin version only matters for this one mojo:

```bash
mvn test-compile
mvn -Dhpi.version=… -Djenkins.version=… -DoverrideWar=… hpi:resolve-test-dependencies surefire:test
```

> then PCT can be nuked from orbit

I think we can continue to use PCT to arrange for checkouts of plugins in the megawar, running Maven, collecting results, etc. The hope is that we can suppress the most complex and buggy portions that deal with POM rewrites. (Also making it less stressful to analyze PCT failures locally. Currently you have to actually run PCT and then open the patched repo clone in an IDE to study test failures—and if you come up with local patches to solve the failure, figure out how to apply those to a clean checkout.)

The handling of multimodule repositories is also unnecessarily complex in PCT. Would be better to aggregate all plugins coming from one repository:

```bash
cd pipeline-model-definition-plugin
mvn -Pquick-build install
mvn -Djenkins.version=… -DoverrideWar=… hpi:resolve-test-dependencies surefire:test
```

which would again be a bit smoother with automatic `jenkins-bom` handling:

```bash
cd pipeline-model-definition-plugin
mvn -DoverrideWar=… install
```

**Comment by basil** [comment]
[As of PR build #36](https://ci.jenkins.io/job/Tools/job/bom/job/PR-1059/36/) I have the first passing `jenkinsci/bom` run with `useUpperBounds`, including a passing `InjectedTest` on all plugins.

**Review by jtnord — APPROVED** [review]

**Comment by jetersen** [comment]
hopefully https://github.com/jenkinsci/bom/pull/1275 have given a good use case for this? 😅

**Comment by basil** [comment]
> hopefully [jenkinsci/bom#1275](https://github.com/jenkinsci/bom/pull/1275) have given a good use case for this? sweat_smile

Sure, but I see little hope of ever completing this project so long as `plugin-compat-tester` maintainers continue to be unresponsive and do not merge my PRs. Help would certainly be appreciated if you are interested.

**Comment by jetersen** [comment]
included in plugin pom https://github.com/jenkinsci/plugin-pom/releases/tag/plugin-4.44 👏

**Pull request merged**
